mypy/plugins/default.py

```diff
@@ -417,3 +419,17 @@ def int_pow_callback(ctx: MethodContext) -> Type:
         else:
             return ctx.api.named_generic_type('builtins.float', [])
     return ctx.default_return_type
+
+
+def int_neg_callback(ctx: MethodContext) -> Type:
+    """Infer a more precise return type for int.__neg__.
+
+    This is mainly used to infer the return type as LiteralType
+    if the original underlying object is a LiteralType object
+    """
+    if isinstance(ctx.type, Instance) and isinstance(ctx.type.last_known_value, LiteralType):
+        value = ctx.type.last_known_value.value
+        fallback = ctx.type.last_known_value.fallback
+        if isinstance(value, int):
+            return LiteralType(-value, fallback)
```

Comment:
You should return an `Instance` here with the last known value set to this literal, instead of the literal directly. This is what screws type inference and makes many tests fail.

Reply:
I summary the message as below:
- add `type_context` to `CheckerPluginInterface`
- fill the missing logic so that `TypeChecker ` correctly has a `type_context`
- copy logic from `ExpressionChecker.infer_literal_expr_type()` to complete this task
If my understanding is correct, I will start working on it ASAP

Reply:
Yes.

(about the last bullet, copy the logic to your plugin)

Reply:
It's my misunderstanding. Now I think it's going to work, please check 32a99a2